### PR TITLE
Check Facility Geocoordinates to make sure they are in Georgia

### LINKED ISSUE #182: Several facilities have incorrect geocoordinates
Sometimes longitude is entered as a positive number. For example, (33.9, 84.3) instead of (33.9, -84.3), which puts it in the wrong hemisphere. Other times, the coordinates are not in Georgia; for example, (38.9, -95.2).

There's not much you can do about typos, but maybe a max/min could be enforced to at least prevent obviously incorrect coordinates. (Allowance should be made for zeros, since that's what is used when coordinates are unknown.)

### PR DESCRIPTION
**Pull request by tom-karasch**
Check to make sure Facility Geocoordinates are within State of Georgia.
First tried modifying Facility update in the repository, but it only shows an error to the user with no feedback. So, added to the Add and Edit Facility pages to give proper feedback to the user if invalid Latitude or Longitude are entered. 

closes: #182

### CODE REVIEW

**Comment by tom-karasch** [comment]
Doug,
I see that SonarCloud does not like duplicated code. Should I create a new "Helper" file to do this GeoCoord. test? If so, it makes it difficult to do the "ModelState.AddModelError" message for the user. 
I'm not sure of the best approach for "Best Practice".

**Comment by tom-karasch** [comment]
**### Doug, Do not review this yet. I am making changes so checks will pass. Thanks**

**Comment by tom-karasch** [comment]
Doug, 
Two tests in FacilityAddTests.cs fail. I spent a while trying to figure out why, but I am not that good at Testing code yet.

File: FMS/Helpers/GeoCoordHelper.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+﻿namespace FMS
+{
+    public static class GeoCoordHelper
```
Comment (dougwaldron) [inline FMS/Helpers/GeoCoordHelper.cs:3]: Style notes: Constant names should be capitalized. Method params should be lowercase. The constants in this class can be private.

**Review by dougwaldron — CHANGES_REQUESTED** [review]
I think your approach -- adding the validation checks in the Add/Edit pages and putting the logic in a helper file -- is the right one. I would adjust it, though, so that zero is only allowed if entered for both latitude and longitude.

I think you can fix the unit tests. When you run the tests in the VS test explorer, it tells you exactly what failed with line numbers. Plus, you can set breakpoints in a unit test and step through it in debug mode. If you want to go through it together, let me know.

Also, please add unit tests for the GeoCoordHelper methods. There's not much to test here, so these will be very straightforward.

**Comment by tom-karasch** [comment]
Still working on the tests

**Review by dougwaldron — APPROVED** [review]
Hey Tom, I went ahead and made a few minor changes rather than send them back around again. A few concepts to keep in mind:

1. Avoid negatively named functions. `!IsValid()` is easier to understand than `IsInvalid()`, especially when writing tests!

2. Sometimes when a test is failing, it's the test that needs to be fixed, not the code. Adding a null check in the Add page fixed the test, but since the framework prevents nulls from being submitted, it's unnecessary code (which will be harder for future developers to understand). A tricky thing about unit tests is that they don't go through the framework validation, so in this case, setting the latitude and longitude to non-null values in the unit test itself is all that's needed.

I think the rest of the questions should be self-explanatory, but let me know if you have any questions on what I changed.

**Comment by tom-karasch** [comment]
> 

Hey Doug,
Thanks for the feedback. 


> 1. Avoid negatively named functions. `!IsValid()` is easier to understand than `IsInvalid()`, especially when writing tests!

**Tom:** Yes, this makes sense. 

> 2. Sometimes when a test is failing, it's the test that needs to be fixed, not the code. Adding a null check in the Add page fixed the test, but since the framework prevents nulls from being submitted, it's unnecessary code (which will be harder for future developers to understand). A tricky thing about unit tests is that they don't go through the framework validation, so in this case, setting the latitude and longitude to non-null values in the unit test itself is all that's needed.

**Tom:** The first thing I noticed were the null values in the test for Lat and Long. Then I saw that the test was used for both Add and Edit, and only failed for Add. And I thought, if I change the test, Doug will say, "Don't change the test to make the code work." So, I added the null check in the Add form, knowing that it was redundant, since the form validation would not allow a null. And I also noticed that the tests don't go through form validation. So, I added the null check in the Add form, but figured you would want to do that a better way, i.e. **fix the test**. LOL.  So, we are on the same page thought process-wise. 
Also, I misunderstood when you said constants should be capitalized. I thought you meant the whole name. Also, I see you made the constants public again so you could use them in the AddModelError messages. This was my intention until you said to make them private. hahaha 👍 

> 
> I think the rest of the questions should be self-explanatory, but let me know if you have any questions on what I changed.

**Tom:** Yes, the rest is self-explanatory. I think we had the same thing in mind, just that my logic was a bit inverted and a bit of miscommunication. 
I also saw that you nested the three checks in an If-Else statement in the pages. It doesn't change the logic, but I see that it could save steps in some cases, for efficiency sake. 
All-in-all, thanks for the feedback and help.

**Comment by dougwaldron** [comment]
Thanks, Tom. I hope this was useful and maybe even enjoyable to work on! Writing unit tests is a very different style of coding and there's an art to it. It's been a fun journey for me and I feel like I'm just starting to get the hang of it.

A failing test always raises the question of whether to fix the code or the test, and it helps to take a step back and ask what is this actually supposed to be testing? In this case, the name of the tests was the clue. We're expecting a certain result *given a valid model.* But the tests were failing because the model wasn't validating, so that's what needed to be fixed. (There are other tests that test what happens when the model *doesn't* validate.)

**Comment by tom-karasch** [comment]
Although I understand perfectly why to make the changes to the Test function logic, I do have a concern still. Now, the tests only use valid Geocoordinates. The reason I have a question about this is that the function is to notify the user of Invalid coordinates. If only valid coordinates are used in the unit tests, then the logic to inform the user of invalid values is not tested. I don't think it is any cause for alarm, but it is something to think about in the future in other projects.

**Pull request merged**